### PR TITLE
Refactor/247 version from source

### LINKED ISSUE #247: Detect project version from source
Currently, PK detects the version in multi maven module projects from the aggregator. However, usually that version is 0.0.0

### PR DESCRIPTION
**Pull request by jakobbraun**
Closes #247

### CODE REVIEW

File: project-keeper/src/main/java/com/exasol/projectkeeper/ProjectVersionDetector.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            }
+        }
+
+        private IllegalArgumentException getSourceHasNoVersionException(final Path requestedPath) {
+            return new IllegalArgumentException(ExaError.messageBuilder("E-PK-CORE-115")
+                    .message(
+                            FAILED_TO_DETECT_VERSION
+                                    + " The specified source with path {{path}} did not provide a version.",
+                            requestedPath)
+                    .mitigation("Please specify a different source to read from or set an explicit version.")
```
Comment (kaklakariada) [inline project-keeper/src/main/java/com/exasol/projectkeeper/ProjectVersionDetector.java:72]:
```suggestion
                    .mitigation("Please specify a different source to read from or set an explicit version in your project-keeper config.")
```

File: project-keeper/src/main/java/com/exasol/projectkeeper/ProjectVersionDetector.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+        private IllegalArgumentException getNoSourceFoundException(final Path requestedPath,
+                final List<AnalyzedSource> analyzedSources) {
+            final List<Path> knownSources = analyzedSources.stream().map(AnalyzedSource::getPath)
+                    .collect(Collectors.toList());
+            return new IllegalArgumentException(ExaError.messageBuilder("E-PK-CORE-114")
+                    .message(FAILED_TO_DETECT_VERSION + " Could not find a source with specified path {{path}}.",
+                            requestedPath.toString())
+                    .mitigation(
+                            "Please make sure that you defined a source with exactly the same path. There following sources are defined in the config: {{sources}}.",
```
Comment (kaklakariada) [inline project-keeper/src/main/java/com/exasol/projectkeeper/ProjectVersionDetector.java:84]:
```suggestion
                            "Please make sure that you defined a source with exactly the same path. The following sources are defined in the config: {{sources}}.",
```

File: project-keeper/src/main/java/com/exasol/projectkeeper/config/ProjectKeeperConfig.java [hunk 3]
```diff
@@ -33,6 +34,65 @@
         MAVEN
     }
 
+    /**
+     * interface for classes that tell PK how to find out the overall project version.
```
Comment (kaklakariada) [inline project-keeper/src/main/java/com/exasol/projectkeeper/config/ProjectKeeperConfig.java:38]:
```suggestion
     * Interface for classes that tell PK how to find out the overall project version.
```

File: project-keeper/src/main/java/com/exasol/projectkeeper/config/ProjectKeeperConfig.java [hunk 4]
```diff
@@ -33,6 +34,65 @@
...
+     */
+    public interface VersionProvider {
```
Comment (kaklakariada) [inline project-keeper/src/main/java/com/exasol/projectkeeper/config/ProjectKeeperConfig.java:40]: This name is misleading. From a `VersionProvider` i would expect a `get()` method that returns the version.
Maybe rename to `VersionConfig`?

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: .project-keeper.yml [hunk 5]
```diff
@@ -37,9 +37,13 @@ sources:
     path: maven-project-crawler/pom.xml
     modules:
       - maven_central
+      - udf_coverage
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply .project-keeper.yml:40]: yes. It's called udf coverage but means: "Get coverage from a JVM that is run during tests." Here we start mvn in the tests

**Review by kaklakariada — APPROVED** [review]

**Pull request merged**